---
title: "Differential gene expression data formats converter"
author: "Andrzej Oleś"
date: "`r doc_date()`"
package: "`r pkg_ver('DEFormats')`"
vignette: >
  %\VignetteIndexEntry{Differential gene expression data formats converter}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
abstract: >
  `r Rpackage("DEFormats")` facilitates data conversion between most widely used 
  tools for differential gene expression analysis; currently, the Bioconductor 
  packages `r Biocpkg("DESeq2")` and `r Biocpkg("edgeR")` are supported.
  It provides converter functions of the form `as.<class>(object)`, where 
  `<class>` is the name of the class to which `object` should be coerced. 
  Additionally, `r Rpackage("DEFormats")` extends the original packages by 
  offering constructors for common count data input formats, e.g. 
  *SummarizedExperiment* objects.
output: 
  BiocStyle::html_document
---

# Convert between *DESeqDataSet* and *DGEList* objects

The following examples demonstrate how to shuttle data between *DESeqDataSet* 
and *DGEList* containers. To begin with, we first load the required libraries.

```{r library, message=FALSE}
library(DESeq2)
library(edgeR)
library(DEFormats)
```

## *DGEList* to *DESeqDataSet*

Let’s create a sample expression data set of an RNA-seq experiment. We will use 
the function provided by `r Rpackage("DEFormats")` to generate a counts table.

```{r counts}
counts = simulateRnaSeqData()
```

The resulting object is a matrix with rows corresponding to genomic features and 
columns to samples.

```{r headcounts}
head(counts)
```

In order to construct a `DGEList` object we additionally need the sample 
grouping.

```{r dge}
group = rep(c("A", "B"), each = 3)

dge = DGEList(counts, group = group)
dge
```

Once we have a `DGEList` object, we can easily convert it to a `DESeqDataSet` 
object with the help of the function `as.DESeqDataSet`.

```{r dds}
dds = as.DESeqDataSet(dge)
dds
```

To make sure that the coercions conserve the data, we can convert `dds` back to 
a `DGEList` and compare it with the original `dge` object.

```{r dgedds}
identical(dge, as.DGEList(dds))
```

Note that because of the use of reference classes in the `SummarizedExperiment`
class which `DESeqDataSet` extends, `identical` will return `FALSE` for any two
`DESeqDataSet` instances, even for ones constructed from the same input:

```{r identicalDDS}
dds1 = DESeqDataSetFromMatrix(counts, data.frame(condition=group), ~ condition)
dds2 = DESeqDataSetFromMatrix(counts, data.frame(condition=group), ~ condition)

identical(dds1, dds2)
```

## *DESeqDataSet* to *DGEList*

Apart from a count matrix, the `simulateRnaSeqData` can also return an 
annotated `RangedSummarizedExperiment` object. The advantage of such an object 
is that, apart from the counts matrix stored in the `assay` slot, it also 
contains sample description in `colData` and gene information stored in 
`rowRanges` as a s`GRanges` object .

```{r se}
se = simulateRnaSeqData(output = "RangedSummarizedExperiment")
se
```

The `se` object can be readily used to construct a `DESeqDataSet` object,

```{r dds_se}
dds = DESeqDataSet(se, design = ~ condition)
```

which can be converted to a `DGEList` object using the familiar method.

```{r dgedds2}
dge = as.DGEList(dds)
dge
```

Please note the additional `genes` element on the `dge` list.

# Creating *DGEList* objects from *SummarizedExperiment*

Similarly as for the *DESeqDataSet* constructor from `r Biocpkg("DESeq2")`,
we can directly use the generic *DGEList* constructor defined in 
`r Rpackage("DEFormats")` on `RangedSummarizedExperiment` objects to create a 
corresponding object in the `r Biocpkg("edgeR")` pipeline:

```{r dgeSE}
dge = DGEList(se, group = colData(se)$condition)
dge
```
